### PR TITLE
Add EventHubs diagnostic instrumentation

### PR DESCRIPTION
**Pull request by pakrym**
Convention based sender spans and activity Id flow.

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs]: What's this thing?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventInstrumentation.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventInstrumentation.cs]: We should consider renaming this to `EventDataInstrumentation` to match the type that it is instrumenting.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs]: This should replace the existing tests that were brought forward to mirror the Track One diagnostics, right?  It's not intended to supplement?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: Should we consider `nameof` for this or do we want this to remain static even if we refactor?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: We should consider doing a `ToList` or some other transform, since we're now enumerating this twice and have no guarantee that it is safe to do so, no?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: :(

**Review by jsquire — COMMENTED** [review]
I've got a feeling that you're still iterating, so I'm going to wait on approval until you indicate that you're ready.

Since you're kindly jumping in and saving me from figuring out how to integrate with things, I'll avoid asking for the missing doc comment blocks and handle them once you merge.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: This we need to figure out how to remove a dependency on.   This class should be agnostic to the transport and not bound to AMQP.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: nit: extra space.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: I get the feeling that we'll see these keys elsewhere; may be worth considering hoisting them into the diagnostics class as `const` members.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: I may be overlooking it, but I don't see where the AMQP reference comes into play at all?  Is this just a leftover from working through things?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: Since we're only really interested in the side effect for the events, maybe it would make sense for a minor abuse of LINQ to avoid having to talk the set twice.   Something like:

```c#
var activity = Activity.Current;
var instrumentedEvents = events.Select(eventData => 
{
      EventInstrumentation.InstrumentEvent(eventData, activity);
      return eventData;
});

//...
await InnerProducer.SendAsync(instrumentedEvents , options, cancellationToken);
// ...
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs]: You have to have `InternalsVisibleTo` to be able to use `Moq` with internal types.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs]: Correct.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Properties/AssemblyInfo.cs]: I... did not know that this was even possible.   That would have saved me a bunch of boilerplate.    I'm stealing this in the future!

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: Was a leftover from some experimentation.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: I don't think mutating as part of enumeration is appropriate here as we don't know what happens with enumerable downrange. The track 1 client enumerates events multiple times...

**Comment by pakrym** [comment]
> missing doc comment 

I've added a parameter comment, something else I've missed?

**Comment by jsquire** [comment]
> > missing doc comment
> 
> I've added a parameter comment, something else I've missed?

Event Hubs has everything fully commented, public and non-public in order to help make it accessible for community contributors and provide context for developers working in the code.   Seriously, though... I don't mind following up and working through it;  its a good forcing function to make me slow down and understand what you did at a deeper level, and it's far less of a time investment than I'd have spent doing the entire implementation, so... don't feel obligated to conform,

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: The track one client is a bad model.   We're accepting an unknown enumeration in a public API that may or may not be safe to walk multiple times.   Generally, my understanding is that unless you're in control of the type and know for sure that it is safe, you want to either do a straight transform to a type that is safe or use the deferred translation to run during enumeration to avoid multiples.   This is, of course, abusing the deferred map for a side effect, but it would potentially save doing 3 walks through the set.   That said, we know the set is not of a large size, so you could make the case that just doing a `ToList` and then using what we have is probably as good, if not potentially better, than adding the LINQ into it.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsTests.cs]: Good deal, I'll make them go away after this gets in, then.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: I've added `ToList()`, we can rethink if it becomes a problem.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubProducer/EventHubProducerTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubProducer/EventHubProducerTests.cs]: Can we please stick with the `Assert.That` for mat.   I believe that the `Is.EquivilantTo` constraint is what you're looking for here.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubProducer/EventHubProducerTests.cs]: Thank you!

Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/EventHubProducer/EventHubProducerTests.cs]: That's what it was!

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs
Comment (lmolkova) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: oh, this would mean that all messages in the batch share the same Id. This does not look right - how would you distinguish those? i.e. causation is broken

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (lmolkova) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: same here - all events have the same Id - causation is lost

**Review by lmolkova — CHANGES_REQUESTED** [review]
still argue that links are needed as each message must carry unique context rather than Id of the incoming request (for EventDataBatch) or Id of Send operation.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: The batch is sent as a single message over the wire, and received by Event Hubs as a single entity.   Each event in that batch is a member, not a stand-alone message.  I'm missing enough overall context around the distributed tracing goals that I'm not able to reconcile whether that addresses your concern or not, but it feels like it would be relevant.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: Likewise, the same additional context as above.  Each event in the batch is sent as a single AMQP message to the Event Hubs service and treated as such.

**Review by jsquire — APPROVED** [review]
From the Event Hubs client point of view, I'm good.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: > oh, this would mean that all messages in the batch share the same Id. 

Not exactly, every message would have the ID of the Activity it was added within.  I'm not sure how adding a different ID to every message helps causation. Do you have some use case in mind?

Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: I think I understand, because there are no other effective message IDs, diagnostic ID is only thing that can uniquly identify the message down the line.

Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: But on the other hand even the doc says ` 'Diagnostic-Id' uniquely identifies operation that enqueued the event `. Which doesn't require ID per message.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: Would one of you be so kind as to define `message` in this context?   It may help me better understand the goal.

Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: When I say message I mean event.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: Is the underlying goal to trace units that correspond to each service interaction or the components of those?   For example, if this were a REST service where we made a POST request, would we be trying to instrument the POST or each item in the payload of that request?

Comment (lmolkova) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: @jsquire  the underlying goal is to trace each event (or samples) end-to-end across multiple services. We will have instrumentation for sending and processing (receiving is a different beast and we may tackle it later).

For HTTP we trace all http requests and each of them carries a unique context.

Event through in case of EventHub we batch event together, their origin (and context they-we been created in) is more important than the fact they-ve been sent together.

Here is the more broad spec on this: https://gist.github.com/lmolkova/e4215c0f44a49ef824983382762e6b92

@pakrym so the spec says
> Every message MUST carry trace context and trace context MUST be unique per message if tracing is enabled

> SDK MUST generate context for message when creating message OR when sending it.

> SDK MUST create a span for each send attempt and add links to each message's context (when send span is not related to message context). SDK MUST set links before starting the span.

This way we create Activity/span per message and create unique id). So when there are two messages that are created in scope of one request, you can distinguish them and trace them separately (causation is preserved by them having unqiue context). If they share the same context you are no longer able to tell which one went where and how it was processed.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs
Comment (lmolkova) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducer.cs]: right, but transport of these events is less of the concern than tracing individual events. It's interesting when a particular event has been sent, but only along with everything that happened with this event before/after.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs
Comment (pakrym) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: Creating a 0 length span/Activity for the message just to generate an ID feels completely wrong. The message is not an operation it's an object.

Comment (lmolkova) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventDataBatch.cs]: you also need to log it - not just generate a unique id. And while it might feel strange, this is the only feasible way to go using existing things we have.

We kinda want timestamp, id (and eventually the rest of the context) and we want to log it somewhere, we may even want tags (message-id or other properties). The two things we don't need for it are duration and status.

So I'd argue it's a reasonable thing to do.

Everyone else does it (google in their pub-sub), zipkin recommends for kafka it and community-based kafka instrumentations do it too.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventDataInstrumentation.cs
Comment (lmolkova) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Diagnostics/EventDataInstrumentation.cs]: i think it's not needed, internal  default in opentelemetry and is a good default in general case

**Review by lmolkova — APPROVED** [review]

**Pull request merged**
